test/e2e/framework/framework.go

```diff
@@ -144,10 +145,19 @@ type Options struct {
 	GroupVersion *schema.GroupVersion
 }
 
-// NewFrameworkWithCustomTimeouts makes a framework with with custom timeouts.
+// NewFrameworkWithCustomTimeouts makes a framework with custom timeouts.
+// For timeout values that are zero the normal default value continues to
+// be used.
 func NewFrameworkWithCustomTimeouts(baseName string, timeouts *TimeoutContext) *Framework {
 	f := NewDefaultFramework(baseName)
-	f.Timeouts = timeouts
+	in := reflect.ValueOf(timeouts).Elem()
+	out := reflect.ValueOf(f.Timeouts).Elem()
+	for i := 0; i < in.NumField(); i++ {
+		value := in.Field(i)
+		if !value.IsZero() {
+			out.Field(i).Set(value)
+		}
+	}
```

Comment:
Why the above change is necessary?
Callers of `framework.NewFrameworkWithCustomTimeouts` expect the timeout values should be replaced with the specified timeout values for each driver.
and this pull request doesn't change those caller sides.

Reply:
Suppose some code uses this function like this:

```
myTimeouts := framework.TimeoutContext{
    // Initialize all fields here.
    PodStart: ...
    ...
}
f := framework.NewFrameworkWithCustomTimeouts("foo", myTimeouts)
```

The API allows that, so we can't be sure that it isn't done. Now we add some new timeout field. With the previous `	f.Timeouts = timeouts`, the tests would run with 0 as value for that new field. With the new code, that field will have a sane default.

The alternative is to not let users create TimeoutContext structs (API change) or require that they obtain one from NewTimeoutContext[FromDefaults] and then change some field (hard to enforce).

This solution seems simpler.

Reply:
Another way to explain this change is this question:

Which is better, a `NewFrameworkWithCustomTimeouts` where the caller must set all fields or a `NewFrameworkWithCustomTimeouts` where the caller only needs to set those fields it cares about?

The latter is easier to use and makes this function usable without calling `NewTimeoutContext` first - perhaps I should make that change. We might even remove that function entirely. I can see some value in having it, but nothing in our code base will use it.

Reply:
I was thinking of this patch:
```patch
diff --git a/test/e2e/storage/drivers/in_tree.go b/test/e2e/storage/drivers/in_tree.go
index 04c4e08285e..51468fb040c 100644
--- a/test/e2e/storage/drivers/in_tree.go
+++ b/test/e2e/storage/drivers/in_tree.go
@@ -1967,9 +1967,10 @@ func (v *azureFileVolume) DeleteVolume(ctx context.Context) {
 }
 
 func (a *azureDiskDriver) GetTimeouts() *framework.TimeoutContext {
-	timeouts := framework.NewTimeoutContext()
-	timeouts.PodStart = time.Minute * 15
-	timeouts.PodDelete = time.Minute * 15
-	timeouts.PVDelete = time.Minute * 20
+	timeouts := &framework.TimeoutContext{
+		PodStart:  time.Minute * 15,
+		PodDelete: time.Minute * 15,
+		PVDelete:  time.Minute * 20,
+	}
 	return timeouts
 }
diff --git a/test/e2e/storage/external/external.go b/test/e2e/storage/external/external.go
index 7150a3d73b8..20c33615d6f 100644
--- a/test/e2e/storage/external/external.go
+++ b/test/e2e/storage/external/external.go
@@ -311,7 +311,7 @@ func (d *driverDefinition) GetDynamicProvisionStorageClass(ctx context.Context,
 }
 
 func (d *driverDefinition) GetTimeouts() *framework.TimeoutContext {
-	timeouts := framework.NewTimeoutContext()
+	timeouts := &framework.TimeoutContext{}
 	if d.Timeouts == nil {
 		return timeouts
 	}
diff --git a/test/e2e/storage/framework/testdriver.go b/test/e2e/storage/framework/testdriver.go
index 6614e8d0280..7910c780708 100644
--- a/test/e2e/storage/framework/testdriver.go
+++ b/test/e2e/storage/framework/testdriver.go
@@ -141,7 +141,7 @@ func GetDriverTimeouts(driver TestDriver) *framework.TimeoutContext {
 	if d, ok := driver.(CustomTimeoutsTestDriver); ok {
 		return d.GetTimeouts()
 	}
-	return framework.NewTimeoutContext()
+	return &framework.TimeoutContext{}
 }
 
 // Capability represents a feature that a volume plugin supports
```

It would be sufficient for using the resulting struct with `NewFrameworkWithCustomTimeouts`. However, there might be other usages of the struct, so let's keep GetDriverTimeouts as-is (= returns fully populated struct).

Reply:
Thank you for your explanation @pohly 

> Now we add some new timeout field. With the previous  `f.Timeouts = timeouts`, the tests would run with 0 as value for that new field. With the new code, that field will have a sane default.

I see, that makes sense.
It is nice to specify necessary customized timeouts only from caller side as this pull request does.